test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/RequestBodyLimitTests.cs

```diff
@@ -0,0 +1,408 @@
+            {
+                var response = await SendRequestAsync(address, "Hello World");
+                Assert.Equal(string.Empty, response);
+            }
+        }
+
+        [ConditionalFact]
+        public async Task ContentLengthExceedsLimit_ReadAsync_ThrowsImmidately()
+        {
+            string address;
+            using (Utilities.CreateHttpServer(out address, options => options.MaxRequestBodySize = 10, httpContext =>
+            {
+                var feature = httpContext.Features.Get<IHttpMaxRequestBodySizeFeature>();
+                Assert.NotNull(feature);
+                Assert.False(feature.IsReadOnly);
+                Assert.Equal(11, httpContext.Request.ContentLength);
+                byte[] input = new byte[100];
+                Assert.Throws<IOException>(() => { var t = httpContext.Request.Body.ReadAsync(input, 0, input.Length); });
+                Assert.Throws<IOException>(() => { var t = httpContext.Request.Body.ReadAsync(input, 0, input.Length); });
+                return Task.FromResult(0);
+            }))
+            {
+                var response = await SendRequestAsync(address, "Hello World");
+                Assert.Equal(string.Empty, response);
+            }
+        }
+
+        [ConditionalFact]
+        public async Task ContentLengthExceedsLimit_ReadBeginEnd_ThrowsImmidately()
+        {
+            string address;
+            using (Utilities.CreateHttpServer(out address, options => options.MaxRequestBodySize = 10, httpContext =>
+            {
+                var feature = httpContext.Features.Get<IHttpMaxRequestBodySizeFeature>();
+                Assert.NotNull(feature);
+                Assert.False(feature.IsReadOnly);
+                Assert.Equal(11, httpContext.Request.ContentLength);
+                byte[] input = new byte[100];
+                Assert.Throws<IOException>(() => httpContext.Request.Body.BeginRead(input, 0, input.Length, null, null));
+                Assert.Throws<IOException>(() => httpContext.Request.Body.BeginRead(input, 0, input.Length, null, null));
+                return Task.FromResult(0);
+            }))
+            {
+                var response = await SendRequestAsync(address, "Hello World");
+                Assert.Equal(string.Empty, response);
+            }
+        }
+
+        [ConditionalFact]
+        public async Task ChunkedExceedsLimit_ReadSync_ThrowsAtLimit()
+        {
+            string address;
+            using (Utilities.CreateHttpServer(out address, options => options.MaxRequestBodySize = 10, httpContext =>
+            {
+                var feature = httpContext.Features.Get<IHttpMaxRequestBodySizeFeature>();
+                Assert.NotNull(feature);
+                Assert.False(feature.IsReadOnly);
+                Assert.Null(httpContext.Request.ContentLength);
+                byte[] input = new byte[100];
+                Assert.Throws<IOException>(() => httpContext.Request.Body.Read(input, 0, input.Length));
```

Comment:
Not that I've seen. Data received in the same packet has always been available together.